proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,618 @@
+
+  -- desugared
+  fromBuilder $
+    toBuilder "foo "
+    <> interpolate (f a b) <> toBuilder " bar "
+    <> interpolate (g x)   <> toBuilder " baz "
+    <> interpolate name    <> toBuilder ""
+
+The string literals there will be handled by ``-XOverloadedStrings`` as usual, if enabled.
+
+Static Semantics
+~~~~~~~~~~~~~~~~
+
+Unsolved ``Interpolate a s`` constraints will default ``s`` to the default ``IsString`` type. That is,
+
+#. If ``OverloadedStrings`` + ``ExtendedDefaultRules`` are enabled, and an ``IsString`` default is specified with ``default (...)``, use that
+#. Otherwise, ``s`` defaults to ``String``
+
+Template Haskell
+~~~~~~~~~~~~~~~~
+
+Template Haskell will add a new constructor:
+
+::
+
+  data Exp
+    = ...
+    | InterpolatedStringE [Either Exp String]
+
+Which mimics the ``HsInterpolatedString`` constructor.
+
+Examples
+--------
+
+Examples were tested with `this gist <https://gist.github.com/brandonchinn178/4d35ed189d7018ca34535ac85442790b>`_ (after desugaring the string interpolation).
+
+Multiline strings
+~~~~~~~~~~~~~~~~~
+
+::
+
+  let x = "hello"
+
+  -- original string
+  let str0 =
+        s"""
+        ${x} world
+        world ${x}
+        ${x} world
+        """
+
+  -- resolve multiline string
+  let str1 = s"${x} world\nworld ${x}\n${x} world"
+
+  -- resolve interpolation
+  let str2 = "hello world\nworld hello\nhello world"
+
+Edge cases
+~~~~~~~~~~
+
+The following interpolated string expressions are also valid:
+
+::
+
+  let x = "hello"
+
+  -- seemingly duplicated closing bracket is valid, as the first one closes the expression
+  -- and the second is a character in the string literal
+  s"${x}} world" == "hello} world"
+
+  -- inline type annotation
+  s"a ${1 :: Int} b" == "a 1 b"
+
+  -- multiline expressions: while ugly and should be avoided, valid syntax
+  s"""
+   foo ${drop
+    1
+    x} world
+   """ == "foo ello world"
+
+  -- braces as characters
+  s"foo ${'{'} bar" == "foo { bar"
+
+  -- comments
+  s"foo ${x -- comment} bar" == "foo hello bar"
+
+  -- OverloadedRecordDot
+  let user = User{name = "Alice"}
+  s"foo ${user.name} bar" == "foo Alice bar"
+
+Effect and Interactions
+-----------------------
+
+An existing program containing ``s"..."`` will break when ``-XStringInterpolation`` is enabled. While there's precedent for this (Template Haskell splices make ``$(...)`` different from ``$ (...)``), this is the first instance where whitespace matters for an alphanumeric identifier. But this is not a big deal:
+
+#. It's unlikely for someone to be naming a function as ``s`` in the first place
+#. Easily mitigatable: just add a space, which improves readability anyway
+#. Prefixing string literals like ``s"..."`` is common in other languages: Python, Scala, Javascript/Typescript, etc. so it shouldn't be a big hurdle for newcomers
+
+When ``-XOverloadedStrings`` is enabled, string interpolation can be used for any ``Buildable`` type. Otherwise, it will only ever build Strings.
```

Comment:
I think they have to also always be `IsString`, if interpolated strings always include parts for some empty string literals?

Reply:
Ah right. That's why my initial proposal had the equivalent of `interpolateRaw :: String -> Builder s`. If seems like keeping `toBuilder :: s -> Builder s` would also require the type to have `IsString` in order to use `-XStringInterpolation`. Is that right @sgraf812? Is that what we want?

Reply:
It seems totally reasonable to me. It would be kind of a weird side effect to be able to use string interpolation syntax for things that were `Buildable` but not `IsString`.

Reply:
I agree it would be weird to require `str` in `s" foo " :: str` to be `IsString str`, but not in `s"" :: str` or `s"${x}${y}"`.

So perhaps we should unconditionally require `IsString` on the string type in the presence of `-XOverloadedStrings`. This has the important benefit that we do not have to specify any defaulting rules, because `str` is already defaulted according to the `IsString` rules. (The proposal should mention this in Effects and Interactions.)

There is the question of whether `IsString` should be a superclass of `Interpolate` then.

(A proposal that would be interesting but out of scope here would be to add an additional method `buildString :: String -> Builder s`  to `Interpolate` with the property that `buildString = toBuilder . fromString` and the long-term goal to deprecate `fromString` and `IsString` with this more efficient method. But that does not solve any of the bemourned problems  of `IsString`, I suppose.)